js/style/style.js

```diff
@@ -638,7 +638,16 @@ Style.prototype = util.inherit(Evented, {
         }
 
         var sourceResults = [];
-        for (var id in this.sources) {
+        var sourcesToQuery = {};
+        if (params.layers) {
+            for (var i = 0; i < params.layers.length; i++) {
+                var sourceId = this._layers[params.layers[i]].source;
+                sourcesToQuery[sourceId] = true;
+            }
+        } else {
+            sourcesToQuery = this.sources;
+        }
+        for (var id in sourcesToQuery) {
```

Comment:
I don't think the impact of `Array.some` or equivalent logic will ever be a bottleneck next to `Source#queryRenderedFeatures`. If it becomes one, we will identify it in the profiler and look for optimizations.